### PR TITLE
feat: add support for create, update and createOrUpdate in spanner Accessor

### PR DESCRIPTION
**Pull request by asthamohta**

### CODE REVIEW

File: accessors/spanner/spanner_accessor.go [hunk 1]
```diff
@@ -201,3 +233,211 @@ func (sp *SpannerAccessorImpl) CreateChangeStream(ctx context.Context, adminClie
...
+	}
+	return nil
+}
+
+// CreatesOrUpdatesDatabase updates an existing Spanner database or creates a new one if one does not exist.
+func (sp *SpannerAccessorImpl) CreateOrUpdateDatabase(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI, driver string, conv *internal.Conv, migrationType string) error {
+	dbExists, err := sp.VerifyDb(ctx, adminClient, dbURI)
+	if err != nil {
+		return err
+	}
```
Comment (Deep1998) [inline accessors/spanner/spanner_accessor.go:312]: There seem to be a bunch  of missing coverage mainly on the error flows in this file. We should try to cover these as well.

File: accessors/spanner/spanner_accessor.go [hunk 2]
```diff
@@ -201,3 +233,211 @@ func (sp *SpannerAccessorImpl) CreateChangeStream(ctx context.Context, adminClie
...
+
+	// The schema we send to Spanner excludes comments (since Cloud
+	// Spanner DDL doesn't accept them), and protects table and col names
+	// using backticks (to avoid any issues with Spanner reserved words).
+	fkStmts := conv.SpSchema.GetDDL(ddl.Config{Comments: false, ProtectIds: true, Tables: false, ForeignKeys: true, SpDialect: conv.SpDialect, Source: driver})
+	if len(fkStmts) == 0 {
+		return
+	}
+	if len(fkStmts) > 50 {
+		fmt.Println(`
```
Comment (Deep1998) [inline accessors/spanner/spanner_accessor.go:380]: This is also an opportunity to replace these print statements with zap logger. It'd be an easy win to replace these print statements as we go.

File: accessors/spanner/spanner_accessor.go [hunk 3]
```diff
@@ -47,6 +67,18 @@ type SpannerAccessor interface {
...
 	// Create a change stream with default options.
 	CreateChangeStream(ctx context.Context, adminClient spanneradmin.AdminClient, changeStreamName, dbURI string) error
+	// Create new Database using conv
+	CreateDatabase(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI string, conv *internal.Conv, driver string, migrationType string) error 
+	// Update Database using conv
+	UpdateDatabase(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI string, conv *internal.Conv, driver string) error
+	// Updates an existing Spanner database or create a new one if one does not exist using Conv
+	CreateOrUpdateDatabase(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI, driver string, conv *internal.Conv, migrationType string) error 
+	// Check whether the db exists and if it does, verify if the schema is what we currently support.
+	VerifyDb(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI string) (dbExists bool, err error)
```
Comment (Deep1998) [inline accessors/spanner/spanner_accessor.go:77]: Pls add unit tests for verifydb and validateddl as well.

File: accessors/spanner/spanner_accessor.go [hunk 4]
```diff
@@ -201,3 +233,211 @@ func (sp *SpannerAccessorImpl) CreateChangeStream(ctx context.Context, adminClie
...
+
+// CreatesOrUpdatesDatabase updates an existing Spanner database or creates a new one if one does not exist.
+func (sp *SpannerAccessorImpl) CreateOrUpdateDatabase(ctx context.Context, adminClient spanneradmin.AdminClient, dbURI, driver string, conv *internal.Conv, migrationType string) error {
+	dbExists, err := sp.VerifyDb(ctx, adminClient, dbURI)
+	if err != nil {
+		return err
+	}
+	if !conv.Audit.SkipMetricsPopulation {
+		// Adding migration metadata to the outgoing context.
+		migrationData := metrics.GetMigrationData(conv, driver, constants.SchemaConv)
```
Comment (Deep1998) [inline accessors/spanner/spanner_accessor.go:315]: I dont think Metric population should not happen inside spanner accessor, this should be moved out prior to calling this function. @shreyakhajanchi any repercussions you can think of?

File: accessors/spanner/spanner_accessor.go [hunk 5]
```diff
@@ -25,10 +27,28 @@ import (
...
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/protobuf/proto"
 )
 
+var (
+	// Set the maximum number of concurrent workers during foreign key creation.
+	// This number should not be too high so as to not hit the AdminQuota limit.
+	// AdminQuota limits are mentioned here: https://cloud.google.com/spanner/quotas#administrative_limits
+	// If facing a quota limit error, consider reducing this value.
+	MaxWorkers       = 50
```
[the remark this answers is not in the document]
Comment (asthamohta) [reply accessors/spanner/spanner_accessor.go:48]: Hey Deep, how is this a dataflow config?

File: accessors/spanner/spanner_accessor.go (hunk 3 above)
Comment (asthamohta) [reply accessors/spanner/spanner_accessor.go:77]: I am not mocking verifyDb and validateDdl so the LOC are covered and tested. i will add error scenarios as well so not line is left untested. Is that okay?

File: accessors/spanner/spanner_accessor.go (hunk 4 above)
Comment (shreyakhajanchi) [reply accessors/spanner/spanner_accessor.go:315]: No, there shouldn't be any repercussions of moving this prior to this function call

File: accessors/spanner/spanner_accessor.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Deep1998) [reply accessors/spanner/spanner_accessor.go:48]: Aah my bad, I had it confused with the streaming.go MaxWorkers.

File: accessors/spanner/spanner_accessor.go (hunk 3 above)
Comment (Deep1998) [reply accessors/spanner/spanner_accessor.go:77]: Yeah that should be fine as long as we get coverage.

File: accessors/spanner/spanner_accessor.go [hunk 6]
```diff
@@ -70,7 +98,7 @@ func (sp *SpannerAccessorImpl) CheckExistingDb(ctx context.Context, adminClient
 	for {
 		select {
 		case <-time.After(5 * time.Minute):
-			fmt.Println("WARNING! API call not responding: make sure that spanner api endpoint is configured properly")
+			logger.Log.Debug("WARNING! API call not responding: make sure that spanner api endpoint is configured properly")
```
Comment (Deep1998) [inline accessors/spanner/spanner_accessor.go:101]: nit: This should be  WARN log not a debug

**Review by Deep1998 — APPROVED** [review]
LGTM

**Pull request merged**